Update to BuildKit 0.12

neersighted opened the pull request:
* Supersedes https://github.com/moby/moby/pull/44530
* Supersedes https://github.com/moby/moby/pull/46035

**- What I did**
Update containerd (library) to v1.7 and BuildKit to v0.12.

Make changes to the code to adjust to the new dependency versions, and fix some TODOs that were waiting on these updates. Leave some new TODOs as well, mostly related to the builder-next code and having to re-implement some things instead of re-using them from BuildKit.

**- How I did it**
With help from my IDE, and with assistance from others (@tonistiigi, @crazy-max, @rumpl) where I ran out of skill and/or time.

**- How to verify it**
CI ✅ 

**- Description for the changelog**
* Update buildkit to `v0.12.2`.
* Update containerd to `v1.7.6`.

**- A picture of a cute animal (not mandatory but encouraged)**
![](https://cdn.discordapp.com/attachments/1132712576812318750/1139648380230975508/20230811_215535.jpg)

neersighted committed:
vendor: github.com/containerd/containerd v1.7.6

The DeepEqual ignore required in the daemon tests is a bit ugly, but it
works given the new protoc output.

We also have to ignore lints related to schema1 deprecations; these do
not apply as we must continue to support this schema version.

Signed-off-by: Bjorn Neergaard <bjorn.neergaard@docker.com>

neersighted committed:
builder-next: apply source policies on config resolve

SourcePolicy was accounted for in https://github.com/moby/buildkit/commit/330cf7ae7d9ac983432ab13f8bc9e7d4aff75695

TODO: replace applySourcePolicies with BuildKit's implementation, which
is currently unexported.

Co-authored-by: Tonis Tiigi <tonistiigi@gmail.com>
Signed-off-by: Bjorn Neergaard <bjorn.neergaard@docker.com>

neersighted committed:
builder-next: adopt new wrapped types

With BuildKit 0.12, some existing types are now required to be wrapped
by new types:

* containerd's LeaseManager and ContentStore have to be a
  (namespace-aware) BuildKit type since https://github.com/moby/buildkit/commit/f044e0a9468639559db93fe30ee826ce502ac481
* BuildKit's solver.CacheManager is used instead of
  bboltstorage.CacheKeyStorage since https://github.com/moby/buildkit/commit/2b30693409271c579d55f9a1e256926de89c9edf
* The MaxAge config field is a bkconfig.Duration since https://github.com/moby/buildkit/commit/e06c96274ffe3bd777c202f121af151b36a9f57d

Signed-off-by: Bjorn Neergaard <bjorn.neergaard@docker.com>

neersighted committed:
builder-next: make stub executor generic

The current executor is only tested on Linux, so let's be honest about
that. Stubbing this correctly helps avoid incorrectly trying to call
into Linux-only code in e.g. libnetwork.

Signed-off-by: Bjorn Neergaard <bjorn.neergaard@docker.com>

neersighted committed:
builder-next/pull: use ResolveMode.String() helper

Introduced years ago in https://github.com/moby/buildkit/commit/6644e1b0dab8cae614207083ec1ba523a495fb44

Signed-off-by: Bjorn Neergaard <bjorn.neergaard@docker.com>

neersighted committed:
builder-next/wrapper: use new exptypes constants

Introduced in https://github.com/moby/buildkit/commit/4fc2d7b5e7ffa51235a30f55432d1f0eb619f77c

Signed-off-by: Bjorn Neergaard <bjorn.neergaard@docker.com>

neersighted committed:
daemon/c8d: use new containerd LabelDistributionSource constant

Introduced in https://github.com/containerd/containerd/commit/dd3eedf3c35c1213292da7c6ef5e033d2a5b5d4d

Signed-off-by: Bjorn Neergaard <bjorn.neergaard@docker.com>

neersighted commented:
Rebased on top of https://github.com/moby/moby/pull/46027

There are still three major issues with this PR:
* Timeouts in BK integration tests (evidently, BK saw something like this with buildkitd and it was related to recent changes they made)
* Tests do not start on Windows (something is trying to reference `/proc`, need to track it down)
* Sometimes containerd leases are missing (also something familiar to the BK team)

I'm working on getting a good debug loop for the BK tests, and @crazy-max is planning to look as soon as he has time.

neersighted commented:
Rebased on top of https://github.com/moby/moby/pull/46032

neersighted commented:
Rebased to pick up https://github.com/moby/buildkit/pull/4042

thaJeztah commented on builder/builder-next/exporter/overrides/wrapper.go:
```diff
@@ -30,14 +23,14 @@ func (e *imageExporterMobyWrapper) Resolve(ctx context.Context, exporterAttrs ma
-	exporterAttrs[keyUnpack] = "true"
-	if _, has := exporterAttrs[keyDanglingPrefix]; !has {
-		exporterAttrs[keyDanglingPrefix] = "moby-dangling"
+	exporterAttrs[string(exptypes.OptKeyName)] = strings.Join(reposAndTags, ",")
+	exporterAttrs[string(exptypes.OptKeyUnpack)] = "true"
```
For these I was wondering if it's possible to do the reverse (cast to `ImageExporterOptKey`) ? Haven't looked at this code locally though, just through GitHub's ui

neersighted replied:
The type in the protobuf is `string`; I don't know that we can control the codegen enough to "do the right thing" here, and this is the pattern BK uses. Still, it's really clunky, and if we can invert it I'm totally in favor.

thaJeztah replied:
Ah... gotcha.. didn't realise we're dealing with protobuf here. Was mostly hoping "strong(ish) types" to help discoverability.

neersighted replied:
Yeah, this is just a newtype to ease discoverability, but it doesn't seem to be totally ergonomic even on the BK side due to protobufs:
https://github.com/moby/buildkit/blob/f53c1752113e1f9c0fbcc1d5c8e18a9b13452afc/control/control.go#L334-L342

crazy-max replied:
Maybe `map<string, google.protobuf.Any>` would have work but the contract is already there.

neersighted replied:
Right, my thought (and specific inxperience) is that maybe we can change the protobuf definition to use the newtype, and existing clients with the old generated code will continue to interoperate as it's just a type alias to string. That breaks the Go API a little; but in theory the protobuf is the stable component and the generated Go code can change (ala containerd v1.7).

neersighted replied:
It turns out this is not possible as we have multiple types for the keys (`ExporterOptKey` for all exporters, `ImageExporterOptKey` for the image exporter).

It looks like this is a wart we'll have to live with; in the future when/if we remove Gogo, the SDK will have to change a fair bit anyway; at that point defining some conversion helper (maybe with generics) ought to be better.

thaJeztah commented:
Oh, panic ain't good; do we know what's causing it? @crazy-max

neersighted commented:
@crazy-max is looking into the lease issue, which we agree is likely a result of BK checking for history in the wrong namespace (though we haven't isolated the fault yet).

I'm looking at the panic myself; it should be straightforward by comparison, and seems to be another bug in BK itself.

crazy-max commented:
Was looking at timed out tests: https://github.com/moby/moby/actions/runs/5612518827/job/15206581916?pr=45966

Same nil dereference as docker-py:

```
$ TEST_KEEP_CACHE=1 TEST_DOCKERD=1 TEST_DOCKERD_BINARY=./.dev/dockerd TESTPKGS=./client TESTFLAGS="-v --parallel=1 --timeout=30m --run=/TestClientGatewaySolve/worker=dockerd$" ./hack/test integration
...
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x2f6bb1b]

goroutine 531 [running]:
github.com/moby/buildkit/executor/resources.(*Monitor).RecordNamespace(0x0, {0xc0019daa20, 0x2a}, {{0x7f57ac58d490?, 0xc002190000?}})
        /go/src/github.com/docker/docker/vendor/github.com/moby/buildkit/executor/resources/monitor.go:188 +0x3b
github.com/moby/buildkit/executor/runcexecutor.(*runcExecutor).Run(_, {_, _}, {_, _}, {{0xbca720, 0xc0007bd640}, {0x0, 0x0}, {0x0, ...}, ...}, ...)
        /go/src/github.com/docker/docker/vendor/github.com/moby/buildkit/executor/runcexecutor/executor.go:305 +0x1b65
github.com/moby/buildkit/solver/llbsolver/ops.(*ExecOp).Exec(0xc001784a80, {0xbe9680, 0xc001b92bd0}, {0xbca900, 0xc0004beea0}, {0xc00062d950?, 0x1, 0x1})
        /go/src/github.com/docker/docker/vendor/github.com/moby/buildkit/solver/llbsolver/ops/exec.go:373 +0x128d
github.com/moby/buildkit/solver.(*sharedOp).Exec.func2({0xbe95d8, 0xc0012b19f0})
        /go/src/github.com/docker/docker/vendor/github.com/moby/buildkit/solver/jobs.go:911 +0x548
github.com/moby/buildkit/util/flightcontrol.(*call[...]).run(0x1000000000001)
        /go/src/github.com/docker/docker/vendor/github.com/moby/buildkit/util/flightcontrol/flightcontrol.go:121 +0x62
sync.(*Once).doSlow(0xc00113be60?, 0xc000a259e0?)
        /usr/local/go/src/sync/once.go:74 +0xc2
sync.(*Once).Do(0xc001757550?, 0xc00122a7b8?)
        /usr/local/go/src/sync/once.go:65 +0x1f
created by github.com/moby/buildkit/util/flightcontrol.(*call[...]).wait
        /go/src/github.com/docker/docker/vendor/github.com/moby/buildkit/util/flightcontrol/flightcontrol.go:165 +0x451
```

crazy-max commented:
This one is linked to https://github.com/moby/moby/pull/45153. Will open a follow-up on BuildKit repo: https://github.com/moby/moby/actions/runs/5614024841/job/15211390558#step:10:14812

```
=== CONT  TestIntegration/TestDiffHardlinks/worker=dockerd
    client_test.go:7069: checkAllReleasable: skipping check for exported tars in non-containerd test
    mergediff_test.go:1253: 
        	Error Trace:	/src/client/client_test.go:6916
        	            				/src/client/mergediff_test.go:1253
        	            				/src/util/testutil/integration/run.go:204
        	Error:      	Received unexpected error:
        	            	failed to load LLB: requested feature diffop  has been disabled on the build server: only enabled with containerd image store backend
        	            	github.com/moby/buildkit/util/stack.Enable
        	            		/src/util/stack/stack.go:77
        	            	github.com/moby/buildkit/util/grpcerrors.FromGRPC
        	            		/src/util/grpcerrors/grpcerrors.go:198
        	            	github.com/moby/buildkit/util/grpcerrors.UnaryClientInterceptor
        	            		/src/util/grpcerrors/intercept.go:41
        	            	google.golang.org/grpc.(*ClientConn).Invoke
        	            		/src/vendor/google.golang.org/grpc/call.go:35
        	            	github.com/moby/buildkit/api/services/control.(*controlClient).Solve
        	            		/src/api/services/control/control.pb.go:2208
        	            	github.com/moby/buildkit/client.(*Client).solve.func2
        	            		/src/client/solve.go:258
        	            	golang.org/x/sync/errgroup.(*Group).Go.func1
        	            		/src/vendor/golang.org/x/sync/errgroup/errgroup.go:75
        	            	runtime.goexit
        	            		/usr/local/go/src/runtime/asm_amd64.s:1598
        	            	failed to solve
        	            	github.com/moby/buildkit/client.(*Client).solve.func2
        	            		/src/client/solve.go:273
        	            	golang.org/x/sync/errgroup.(*Group).Go.func1
        	            		/src/vendor/golang.org/x/sync/errgroup/errgroup.go:75
        	            	runtime.goexit
        	            		/usr/local/go/src/runtime/asm_amd64.s:1598
        	Test:       	TestIntegration/TestDiffHardlinks/worker=dockerd
```

thaJeztah commented:
> Some buildkit CI failures;
> 
> ```
> === FAIL: frontend TestFrontendIntegration/TestRefEvaluate/worker=dockerd (1.27s)
>     frontend_test.go:341: 
>         	Error Trace:	/src/frontend/frontend_test.go:341
>         	            				/src/util/testutil/integration/run.go:90
>         	            				/src/util/testutil/integration/run.go:204
>         	Error:      	Received unexpected error:
>         	            	lease does not exist: not found
> ```

I think I may have some (partial) fixes for this on my old PR; https://github.com/moby/moby/pull/44530

But Kevin just mentioned there's a fix for that in protobuf

crazy-max commented:
> But Kevin just mentioned there's a fix for that in protobuf

Ok no fix in protobuf repo after all but recall this one https://github.com/moby/buildkit/issues/3187#issuecomment-1280811885 with same lint issues but don't remember how I fixed this in BuildKit...

Edit: here is the fix done in BuildKit: https://github.com/moby/buildkit/commit/e1b3b6c4abf7684f13e6391e5f7bc9210752687a

crazy-max commented:
@neersighted Missing this commit after rebease: https://github.com/crazy-max/moby/commit/cdcd09db25254d1ca755c4ad06aded57d35c6ebf

neersighted commented:
Oops; reapplied the diff

crazy-max commented:
```
 === FAIL: client TestIntegration/TestBasicS3CacheImportExport/worker=dockerd-containerd (5.81s)
2023/07/21 14:34:31 http2: server connection error from localhost: connection error: PROTOCOL_ERROR
    client_test.go:5055: 
        	Error Trace:	/src/client/client_test.go:5055
        	            				/src/client/client_test.go:5196
        	            				/src/util/testutil/integration/run.go:90
        	            				/src/util/testutil/integration/run.go:204
        	Error:      	Received unexpected error:
        	            	unknown cache exporter: "s3"
```

```
=== FAIL: client TestIntegration/TestBasicAzblobCacheImportExport/worker=dockerd-containerd (5.25s)
    client_test.go:5055: 
        	Error Trace:	/src/client/client_test.go:5055
        	            				/src/client/client_test.go:5233
        	            				/src/util/testutil/integration/run.go:90
        	            				/src/util/testutil/integration/run.go:204
        	Error:      	Received unexpected error:
        	            	unknown cache exporter: "azblob"
```

We need either to handle `s3`, `azblob` in moby or skip tests for specific cache exporters. I think the latter is what most maintainers want.

neersighted commented:
I'd say let's go ahead and skip them.

crazy-max committed:
ci(buildkit): remove regex skipping tests with digest inconsistency

Skipping digest-related tests is no longer necessary after https://github.com/containerd/containerd/commit/40658316526ee5e0ec91d110e5999abb0b9a5407

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>
Signed-off-by: Bjorn Neergaard <bjorn.neergaard@docker.com>

crazy-max commented:
> We need either to handle `s3`, `azblob` in moby or skip tests for specific cache exporters. I think the latter is what most maintainers want.

This is actually a regression since v0.12. Should be fixed by https://github.com/moby/buildkit/pull/4055

neersighted commented:
Rebased to pick up https://github.com/moby/buildkit/pull/4055

neersighted committed:
vendor: github.com/moby/buildkit v0.12.2

The following changes were required:
* integration/build: progressui's signature changed in https://github.com/moby/buildkit/commit/6b8fbed01e7b857e53d5378b04152b583fae670e
* builder-next: flightcontrol.Group has become a generic type in https://github.com/moby/buildkit/commit/8ffc03b8f0426f6426cb25db744dc8976de0dccf
* builder-next/executor: add github.com/moby/buildkit/executor/resources types, necessitated by https://github.com/moby/buildkit/commit/6e87e4b455cdbf5454bb1db2da859a4d586c828b
* builder-next: stub util/network/Namespace.Sample(), necessitated by https://github.com/moby/buildkit/commit/963f16179f04c9061a8ecdb9c5cc813b99e1e423

Co-authored-by: CrazyMax <crazy-max@users.noreply.github.com>
Co-authored-by: Sebastiaan van Stijn <github@gone.nl>
Signed-off-by: Bjorn Neergaard <bjorn.neergaard@docker.com>

crazy-max commented:
Btw since BuildKit 0.12 you can now generate GH annotations when running tests: https://github.com/moby/buildkit/pull/3727 if you want top also display them on moby.

crazy-max committed:
ci(buildkit): expose github runtime for gha tests

This exposes `ACTIONS_RUNTIME_TOKEN` and `ACTIONS_CACHE_URL`, which are
used to skip cache exporter tests, when combined with
https://github.com/moby/buildkit/commit/a8789cbd4a657836d1b747a9c4eb54fed4e221bd

Co-authored-by: Bjorn Neergaard <bjorn.neergaard@docker.com>
Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>
Signed-off-by: Bjorn Neergaard <bjorn.neergaard@docker.com>

crazy-max commented:
Rebased and push extra commit to fix gha cache exporter tests:

```
=== FAIL: client TestIntegration/TestDiffSelf/worker=dockerd-containerd (1.33s)
    client_test.go:7069: checkAllReleasable: skipping check for exported tars in non-containerd test
    mergediff_test.go:1253: 
        	Error Trace:	/src/client/client_test.go:6916
        	            				/src/client/mergediff_test.go:1253
        	            				/src/util/testutil/integration/run.go:204
        	Error:      	Received unexpected error:
        	            	failed to load cache key: failed to authorize: failed to fetch anonymous token: unexpected status from GET request to https://auth.docker.io/token?scope=repository%3Alibrary%2Falpine%3Apull&service=registry.docker.io: 503 Service Unavailable
        	            	github.com/moby/buildkit/util/stack.Enable
```

crazy-max commented:
Thanks @tonistiigi for this fix with leaseutil!

Only two tests lefts to fix when testing from my fork: https://github.com/crazy-max/moby/actions/runs/5809077166/job/15747295958#step:11:2523

```
=== RUN   TestIntegration/TestSourcePolicy/worker=dockerd/Frontend_policies/resolve_image_config
    client_test.go:9597: 
        	Error Trace:	/src/client/client_test.go:9597
        	            				/src/frontend/gateway/grpcclient/client.go:214
        	            				/src/client/build.go:59
        	            				/src/client/solve.go:283
        	            				/src/vendor/golang.org/x/sync/errgroup/errgroup.go:75
        	            				/usr/local/go/src/runtime/asm_amd64.s:1598
        	Error:      	Not equal: 
        	            	expected: "docker.io/library/busybox:latest"
        	            	actual  : "docker.io/library/busybox:1.34.1-uclibc"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-docker.io/library/busybox:latest
        	            	+docker.io/library/busybox:1.34.1-uclibc
        	Test:       	TestIntegration/TestSourcePolicy/worker=dockerd/Frontend_policies/resolve_image_config
```

https://github.com/crazy-max/moby/actions/runs/5809077166/job/15747295958#step:11:3827

```
=== FAIL: client TestIntegration/TestSourcePolicy/worker=dockerd/Frontend_policies/resolve_image_config (1.04s)
    client_test.go:9597: 
        	Error Trace:	/src/client/client_test.go:9597
        	            				/src/frontend/gateway/grpcclient/client.go:214
        	            				/src/client/build.go:59
        	            				/src/client/solve.go:283
        	            				/src/vendor/golang.org/x/sync/errgroup/errgroup.go:75
        	            				/usr/local/go/src/runtime/asm_amd64.s:1598
        	Error:      	Not equal: 
        	            	expected: "docker.io/library/busybox:latest"
        	            	actual  : "docker.io/library/busybox:1.34.1-uclibc"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-docker.io/library/busybox:latest
        	            	+docker.io/library/busybox:1.34.1-uclibc
        	Test:       	TestIntegration/TestSourcePolicy/worker=dockerd/Frontend_policies/resolve_image_config
            --- FAIL: TestIntegration/TestSourcePolicy/worker=dockerd/Frontend_policies/resolve_image_config (1.04s)
```

We might need another feature compat check for policies or do the implementation in moby.

crazy-max commented:
BuildKit tests look good but still some issues with Windows tests related to containerd 1.7. Could we vendor containerd 1.7 in another PR to reduce noise here? I see containerd binary has already been bumped to 1.7 but not vendoring :thinking:. Is it expected? I think we still ship containerd 1.6 in our packaging but I see some backports to 24 branch https://github.com/moby/moby/pull/46103?

TBBle commented:
> I see containerd binary has already been bumped to 1.7 but not vendoring 🤔. Is it expected?

Per https://github.com/containerd/containerd/blob/main/RELEASES.md#public-api-stability this should be valid for containerd (unless we're vendoring internal things and relying on them to match the other end of the RPC connection, I guess?), and since the Go Client API is _not_ stable and uses a lot of unstable-API stuff too, it lets us split "Update containerd binary" ("simple") from "Update containerd vendoring", given the latter can be a hassle due to BuildKit or other ~~containerd vendorers...~~ ~~containerd vendees...~~ other things that vendor from or overlap with the vendored things in containerd.

slonopotamus commented:
> ~~containerd vendorers... containerd vendees...~~

containerd dependants! It doesn't matter whether they use vendoring or not.

TBBle commented:
Well, there's a large difference between "Calling the stable GRPC API" and "vendoring from containerd, and hence vendoring from everything that containerd vendors from, and so on" in terms of upgrade hassle, as demonstrated in #44530 and its OpenTelemetry shenanigans.

AkihiroSuda commented:
> It's also worth noting that the static packages/development container are already on containerd 1.7, but we're shipping containerd 1.6 packages on download.docker.com. At least for users of Docker CE, even after this PR, we intend to keep shipping containerd 1.6 binaries for some time, as that is the LTS version.

Would it be possible to create “containerd.io-1.6” and “containerd.io-1.7” packages?
Docker v20.10, v23, and v24 will depend on “containerd.io-1.6” while v25 will depend on “containerd.io-1.7”.
`apt-get install containerd.io` will install 1.7.

neersighted commented:
> Would it be possible to create “containerd.io-1.6” and “containerd.io-1.7” packages?
Docker v20.10, v23, and v24 will depend on “containerd.io-1.6” while v25 will depend on “containerd.io-1.7”.
apt-get install containerd.io will install 1.7.

With Docker CE we really only intend to package the latest branch, so I'm not sure this is something that makes a lot of sense under that model. However, maybe we could add docs callouts suggesting that users of older branches (which are no longer being built) might prefer an older containerd, if that doesn't spawn too much confusion?

neersighted commented:
@rumpl just did an amazing job of tracking down the bad hcsshim change that was breaking us (https://github.com/rumpl/hcsshim/commit/3e090b05a82c0e226e26c0d1e2ae89d336d7ff69).

I've included the patch (plus a workaround for hcsshim main) in this PR, but we should drop them (and re-vendor a fixed hcsshim) before merge/once the fix is accepted upstream (PR pending).

Otherwise, this PR is finally ready for review :tada:

thaJeztah commented:
> @rumpl just did an amazing job of tracking down the bad hcsshim change that was breaking us https://github.com/microsoft/hcsshim/commit/3e090b05a82c0e226e26c0d1e2ae89d336d7ff69
> 
> I've included the patch (plus a workaround for hcsshim main) in this PR, but we should drop them (and re-vendor a fixed hcsshim) before merge/once the fix is accepted upstream (PR pending).
> 
> Otherwise, this PR is finally ready for review 🎉

Was a PR already opened?

/cc @katiewasnothere @kevpar @helsaawy

rumpl commented:
> > @rumpl just did an amazing job of tracking down the bad hcsshim change that was breaking us [microsoft/hcsshim@3e090b0](https://github.com/microsoft/hcsshim/commit/3e090b05a82c0e226e26c0d1e2ae89d336d7ff69)
> > I've included the patch (plus a workaround for hcsshim main) in this PR, but we should drop them (and re-vendor a fixed hcsshim) before merge/once the fix is accepted upstream (PR pending).
> > Otherwise, this PR is finally ready for review 🎉
> 
> Was a PR already opened?
> 
> /cc @katiewasnothere @kevpar @helsaawy

I haven't opened one, will do tomorrow, the fix is https://github.com/rumpl/hcsshim/tree/fix-close-stdin

cpuguy83 commented on builder/builder-next/adapters/containerimage/pull.go:
```diff
@@ -860,3 +872,41 @@ func platformMatches(img *image.Image, p *ocispec.Platform) bool {
+			return "", errors.Errorf("could not parse ref: %s", op.GetSource().GetIdentifier())
+		}
+		if ok && t != srctypes.DockerImageScheme {
+			return "", &imageutil.ResolveToNonImageError{Ref: str, Updated: newRef}
```
This is probably fine as is since it is what is still in buildkit, but this was meant to be  the full identifier.

ref: https://github.com/moby/buildkit/pull/4209

neersighted replied:
Ideally we wouldn't be reinventing this since it's sufficiently abstracted from the snapshotters/graphdrivers, but the implementation in BuildKit isn't really shaped right for us to re-use. I've asked @tonistiigi and he's interested in cleaning this up so we can share it; but assuming that doesn't happen immediately, would you be willing to port your fix once it's accepted in BuildKit?

cpuguy83 reviewed (commented):
Seems mostly straight-forward.
Outside of the DNM commits, this LGTM

AkihiroSuda commented on vendor.mod:
```diff
@@ -190,23 +197,22 @@ require (
 	sigs.k8s.io/yaml v1.3.0 // indirect
 )
+
+replace github.com/Microsoft/hcsshim => github.com/rumpl/hcsshim v0.0.0-20230913144325-3a73df9882a4
```
Is there any pending PR?

rumpl replied:
Yes https://github.com/microsoft/hcsshim/pull/1899

dmcgowan commented on builder/builder-next/worker/worker.go:
```diff
@@ -186,13 +188,13 @@ func (w *Worker) Close() error {
-// ContentStore returns content store
-func (w *Worker) ContentStore() content.Store {
+// ContentStore returns the wrapped content store
+func (w *Worker) ContentStore() *containerdsnapshot.Store {
```
Why is this changed from interface to the pointer type? I couldn't find where it was used differently in this PR

crazy-max commented on builder/builder-next/worker/worker.go:
```diff
@@ -7,9 +7,11 @@ import (
 
+	containerdsnapshot "github.com/moby/buildkit/snapshot/containerd"
+	"github.com/moby/buildkit/util/leaseutil"
+
```
nit: remove newline

slonopotamus replied:
Is there any tool that would automatically format imports properly?

crazy-max replied:
I don't think there is any rule in goimports or depguard to handle this case unfortunately.

neersighted replied:
🙈

crazy-max commented on vendor.mod:
```diff
@@ -87,33 +87,42 @@ require (
 	resenje.org/singleflight v0.4.0
 )
 
+require (
+	google.golang.org/genproto/googleapis/api v0.0.0-20230525234035-dd9d682886f9
+	google.golang.org/protobuf v1.31.0
+)
```
should be merged with "require" block above

neersighted replied:
This is part of the DNM/temporary commit to pull in @rumpl's hcsshim PR.

crazy-max reviewed (commented):
small nits

rumpl commented on builder/builder-next/controller.go:
```diff
@@ -135,7 +135,7 @@ func newSnapshotterController(ctx context.Context, rt http.RoundTripper, opt Opt
 		WorkerController: wc,
 		Frontends:        frontends,
-		CacheKeyStorage:  cacheStorage,
+		CacheManager:     solver.NewCacheManager(context.TODO(), "local", cacheStorage, worker.NewCacheResultStorage(wc)),
```
Any reason we don't use `ctx` here? Same question for the same line in `newGraphDriverController`

neersighted replied:
This is based on buildkitd not passing the context here; cc @crazy-max

neersighted replied:
Looking closer, I don't see any reason **not** to plumb it through, so I've gone ahead and done it 😇

neersighted commented on vendor.mod:
```diff
@@ -210,3 +210,5 @@ require (
 	sigs.k8s.io/yaml v1.3.0 // indirect
 )
+
+replace github.com/Microsoft/hcsshim => github.com/anmaxvl/hcsshim v0.10.0-rc.1.0.20230914180317-064d46e61813
```
This is mildly DNM; still waiting on a merge of https://github.com/microsoft/hcsshim/pull/1900

thaJeztah replied:
this commit was dropped after https://github.com/moby/moby/pull/46522

vvoland reviewed (approved):
LGTM

thaJeztah commented on cmd/dockerd/daemon.go:
```diff
@@ -516,7 +516,7 @@ func loadDaemonCliConfig(opts *daemonOptions) (*config.Config, error) {
 	conf.Hosts = opts.Hosts
 	conf.LogLevel = opts.LogLevel
-	conf.LogFormat = opts.LogFormat
+	conf.LogFormat = log.OutputFormat(opts.LogFormat)
```
Perhaps better to move this to a separate PR, for visibility (the change should not strictly be needed as part of the update now)

thaJeztah replied:
I moved this commit to https://github.com/moby/moby/pull/46525

thaJeztah commented on libcontainerd/remote/client.go:
```diff
@@ -13,6 +13,8 @@ import (
 	"syscall"
 	"time"
 
+	"github.com/opencontainers/go-digest"
+
 	"github.com/containerd/containerd"
```
nit: looks like this ended up in a separate import group.


This commit should probably be combined with the containerd update (as without it, that one won't compile?)

thaJeztah replied:
Fixed this while doing the rebase

thaJeztah commented on vendor.mod:
```diff
@@ -111,6 +111,8 @@ require (
 	resenje.org/singleflight v0.4.0
 )
 
+require google.golang.org/protobuf v1.31.0
+
```
Can you move this one to the first group? (go modules often needs some hand-holding)

thaJeztah replied:
Also guessing that this commit needs to be together with the containerd update?

thaJeztah commented on daemon/runtime_unix_test.go:
```diff
@@ -242,11 +244,11 @@ func TestGetRuntime(t *testing.T) {
 	stockRuntime.Features = nil
 
-	configdOpts := *stockRuntime.Opts.(*v2runcoptions.Options)
+	configdOpts := proto.Clone(stockRuntime.Opts.(*v2runcoptions.Options)).(*v2runcoptions.Options)
```
This commit probably as well together with the containerd update (or does it compile without?)

thaJeztah commented on builder/builder-next/adapters/containerimage/pull.go:
```diff
@@ -19,7 +19,7 @@ import (
 	"github.com/containerd/containerd/remotes"
 	"github.com/containerd/containerd/remotes/docker"
-	"github.com/containerd/containerd/remotes/docker/schema1"
+	"github.com/containerd/containerd/remotes/docker/schema1" //nolint:staticcheck // Ignore SA1019: "github.com/containerd/containerd/remotes/docker/schema1" is deprecated: use images formatted in Docker Image Manifest v2, Schema 2, or OCI Image Spec v1.
```
probably also needs a squash with the containerd update

TBBle replied:
https://github.com/moby/moby/actions/runs/6261172963/job/17000742211?pr=45966 caught another schema1-deprecation that needs lint-skipping.

```
daemon/containerd/image_pull.go:118:22: SA1019: containerd.WithSchema1Conversion is deprecated: use Schema 2 or OCI images. (staticcheck)
	opts = append(opts, containerd.WithSchema1Conversion)
	                    ^
```

I noticed #42300 doesn't appear to remove either of these though, unless I'm misunderstanding, and that's for something else.

neersighted replied:
Thanks for spotting that!

thaJeztah replied:
https://github.com/moby/moby/pull/42300 probably still needs some cleaning up; it started from a much larger branch, then chunks got "chipped off" from that and merged as separate PRs where we could.

thaJeztah commented on integration/build/build_traces_test.go:
```diff
@@ -57,7 +57,7 @@ func TestBuildkitHistoryTracePropagation(t *testing.T) {
 
 	eg.Go(func() error {
-		_, err := progressui.DisplaySolveStatus(ctxGo, "test", nil, &testWriter{t}, ch)
+		_, err := progressui.DisplaySolveStatus(ctxGo, nil, &testWriter{t}, ch, progressui.WithPhase("test"))
```
Should this go into the buildkit update commit?

thaJeztah commented on builder/builder-next/adapters/containerimage/pull.go:
```diff
@@ -92,8 +98,9 @@ func (is *Source) resolveLocal(refStr string) (*image.Image, error) {
 	return img, nil
 }
 
-func (is *Source) resolveRemote(ctx context.Context, ref string, platform *ocispec.Platform, sm *session.Manager, g session.Group) (digest.Digest, []byte, error) {
+func (is *Source) resolveRemote(ctx context.Context, ref string, platform *ocispec.Platform, sm *session.Manager, g session.Group) (string, digest.Digest, []byte, error) {
```
Looks like this commit is needed for the BuildKit update (otherwise it doesn't compile)

neersighted replied:
I'd rather keep this separate, but it is possible to first stub this (like I did the first time), and then apply Tonis's implementation on top. That's not my preference, however, as there will have to be non-buildable commits in this PR to keep things reviewable/discrete.

thaJeztah commented on builder/builder-next/adapters/containerimage/pull.go:
```diff
@@ -198,7 +198,7 @@ func (is *Source) Resolve(ctx context.Context, id source.Identifier, sm *session
 	is               *Source
 	resolveLocalOnce sync.Once
-	g                flightcontrol.Group
+	g                flightcontrol.Group[interface{}]
```
This commit as well (buildkit)

thaJeztah commented on builder/builder-next/executor_linux.go:
```diff
@@ -1,5 +1,3 @@
-//go:build !windows
-
 package buildkit
```
Can you move this commit "before" the buildkit update?

thaJeztah commented on builder/builder-next/executor_linux.go:
```diff
@@ -119,6 +119,11 @@ func (iface *lnInterface) init(c *libnetwork.Controller, n *libnetwork.Network)
 	iface.ep = ep
 }
 
+// TODO(neersighted): Unstub Sample(), and collect data from the libnetwork Endpoint.
+func (iface *lnInterface) Sample() (*network.Sample, error) {
```
> Can you move this commit "before" the buildkit update?

So that this commit can be squashed with the buildkit update (so that the commit compiles)

thaJeztah commented on daemon/config/builder.go:
```diff
@@ -62,8 +63,8 @@ type BuilderGCConfig struct {
 type BuilderHistoryConfig struct {
-	MaxAge     int64 `json:",omitempty"`
-	MaxEntries int64 `json:",omitempty"`
+	MaxAge     bkconfig.Duration `json:",omitempty"`
```
This change does not affect how the config is interpreted? (i.e. existing config still means the same?) because I _think_ this may be part of config options set through `daemon.json` 😅

neersighted replied:
Existing options should work fine; you can see in https://github.com/moby/buildkit/commit/e06c96274ffe3bd777c202f121af151b36a9f57d that a custom deserializer is implemented to handle the existing int64 format, but also parse a Go `time.Duration` string.

thaJeztah replied:
Ah, yes looks good. I didn't look yet, and was curious if the `* times.Second` was handled; I've (ahum) seen cases where that part was overlooked, and now the duration going from "seconds" to "nanoseconds" 😂

thaJeztah reviewed (commented):
changes "LGTM"

- left some nits about imports and "requires" going in the wrong group
- commit-order and squashing
- some minor change that _could_ be moved separate

thaJeztah commented:
@neersighted I did a quick rebase of this one after https://github.com/moby/moby/pull/46522 was merged;

- rebased
- dropped the temporary commit / replace for hcsshim (no longer needed after https://github.com/moby/moby/pull/46522)
- moved the daemon-config log-format type change to a separate PR; https://github.com/moby/moby/pull/46525
- fixed some of the minor nits w.r.t. `vendor.mod` formatting and imports in the wrong "group"

We should probably still squash some of the commits so that the intermediate commits can compile (we can keep some of the "improvements", such as functions that can now be used from buildkit and/or containerd separate if we want them to be more visible in history).

TBBle commented on builder/builder-next/executor_nolinux.go:
If we get a builder-next executor for Windows, will this file need to be renamed _again_? What about `executor_notimplemented.go`, since we're not relying on the GOOS filename tag anymore, and then any future executors only need to update the build-tag to exclude their newly-implemented GOOS from the not-implemented list.

Or am I misunderstanding, and this can never be implemented on non-Linux, e.g., even if we have BuildKit working for Windows.

thaJeztah replied:
The only consideration could be to name this `_unsupported.go` (done that in some other cases, to be slightly more generic, and sometimes preventing some renames).

Not a blocker for me though

AkihiroSuda commented on .github/workflows/buildkit.yml:
```diff
@@ -73,6 +73,9 @@ jobs:
           echo "BUILDKIT_TEST_DISABLE_FEATURES=${disabledFeatures}" >> $GITHUB_ENV
+      -
+        name: Expose GitHub Runtime
+        uses: crazy-max/ghaction-github-runtime@v2
```
Could you add comments to explain why this is necessary?

crazy-max replied:
This will expose `ACTIONS_RUNTIME_TOKEN` and `ACTIONS_RUNTIME_URL` env vars otherwise gha cache integration test is skipped: https://github.com/moby/buildkit/blob/v0.12.2/client/client_test.go#L5407-L5411

@neersighted Can you add this comment? Don't want to jeopardize your branch :sweat_smile:

corhere commented on builder/builder-next/executor_nolinux.go:
```diff
@@ -0,0 +1,34 @@
+type stubExecutor struct{}
+
+func (w *stubExecutor) Run(ctx context.Context, id string, root executor.Mount, mounts []executor.Mount, process executor.ProcessInfo, started chan<- struct{}) (resourcetypes.Recorder, error) {
+	return nil, fmt.Errorf("buildkit executor not implemented for %s", runtime.GOOS)
```
Nit: build the string at compile time as `runtime.GOOS` is a compile-time constant.
```suggestion
	return nil, errors.New("buildkit executor not implemented for "+runtime.GOOS)
```

neersighted committed:
builder-next/a/ci/pull: strongly type flightcontrol.Group

Now that this is a generic, we can define a struct type at the package
level, and remove the casting logic necessary when we had to use
interface{}.

Signed-off-by: Bjorn Neergaard <bjorn.neergaard@docker.com>

thaJeztah reviewed (approved):
LGTM

tianon commented on vendor/github.com/moby/buildkit/solver/result/result.go:
```diff
@@ -1,13 +1,12 @@
 )
 
-type Result[T any] struct {
+type Result[T comparable] struct {
```
```
# github.com/moby/buildkit/frontend/gateway/client
vendor/github.com/moby/buildkit/frontend/gateway/client/client.go:18:29: Reference does not implement comparable
note: module requires Go 1.20
```

If we want to keep this change, we should update `vendor.mod` to `go 1.20`, right?

neersighted replied:
Ah, good catch! I should have tested with an older Go compiler.

tianon replied:
I guess that's https://github.com/moby/moby/pull/46069 👀

thaJeztah replied:
Ah, yes, I decided not to _force_ go1.21 in that PR (as most of the code definitely would still work on older versions); perhaps we still should? At least setting the `go.mod` to go1.20 could be reasonable I guess. (with go1.19 being EOL); at least for the master / v25.0-dev branch.

thaJeztah replied:
I'll make it a separate commit, as we don't need it for the 24.0 (and other branches)

I guess in that case we might as well make it a separate PR; want to open one?

neersighted replied:
https://github.com/moby/moby/pull/46562
